### DISCUSSION TITLE
Asset catalog Color Sets

### DISCUSSION THREAD

**Discussion by johnrbent** [skiptools/skip#202] [Q&A]
Hello!

I see that [asset catalog image sets are supported](https://skip.tools/docs/modules/skip-ui/#image-assets) with light and dark variants. It looks like color sets are not yet supported? Are there any plans to add that?

I did see in this [discussion](https://github.com/orgs/skiptools/discussions/102) that you added `colorScheme` support that could be used to achieve the same effect, but it requires defining each color programmatically and observing `colorScheme` in each view, and writing ternary switches in each modifier. (Maybe there's a simpler way that I'm not aware of)

Mostly I'm just trying to decide the best way to approach building my app with full light/dark variants, which is typically pretty simple in SwiftUI with color sets.

**Comment by johnrbent** [comment]
Well I learned a lot trying to figure out the best approach to this, and finally settled on something I can work with...

The missing piece for me was the `ShapeStyle` protocol. I always overlooked this in all of the SwiftUI styling interfaces and just blindly passed in a `Color` from my color sets that already had light/dark values. But it's perfect for this workaround, and I can just pass in this `CustomColor` type everywhere I would normally pass in a `Color` with the same dot-notation and autocomplete ease. Here's how it's defined:
```
import SwiftUI
#if SKIP
import androidx.compose.foundation.isSystemInDarkTheme
#endif

public struct CustomColor: ShapeStyle, View {
    public let light: Color
    public let dark: Color
    
    #if !SKIP
    public func resolve(in environment: EnvironmentValues) -> some ShapeStyle {
        if environment.colorScheme == .light {
            return light
        } else {
            return dark
        }
    }
    #else
    @Composable public override func asColor(opacity: Double, animationContext: ComposeContext?) -> androidx.compose.ui.graphics.Color? {
        isSystemInDarkTheme() ? dark.colorImpl() : light.colorImpl()
    }
    #endif
}

extension CustomColor {
    public var inverse: CustomColor {
        CustomColor(light: dark, dark: light)
    }
}

extension ShapeStyle where Self == CustomColor {
    public static var primaryText: CustomColor {
        CustomColor(
            light: Color(red: 0.020, green: 0.051, blue: 0.157),
            dark: Color(red: 0.984, green: 0.980, blue: 0.973)
        )
    }
    
    public static var primaryBackground: CustomColor {
        CustomColor(
            light: Color(red: 0.984, green: 0.980, blue: 0.973),
            dark: Color(red: 0.020, green: 0.051, blue: 0.157)
        )
    }
    
    ...
}
```

One caveat is that this only works for iOS 17+ due to use of the `resolve` function. But it's getting to be version-dropping season, so not that big of a deal for most folks. I noticed in the SkipUI source that this function is there but commented out, so maybe this is on your radar...

I also haven't thoroughly tested all of the hierarchical system colors, but they appeared to work in my limited testing (I don't use them in my own projects)

I still think that color sets are the ideal and would reduce a good bit of upfront friction for people like myself who like to start projects this way by setting up the whole UI styling framework so it's in place on every `View` when it comes time to pick a theme (colors, fonts, etc).

**Reply by aabewhite** [reply]
Thanks, this is great! We'd definitely like to get named colors working for everyone. And when we do, perhaps we'll be able to take advantage of some of this code!

**Reply by tomauty** [reply]
Any update on plans to support named colors? Importing quite a few for my current project and realized after the fact that they aren't supported yet :)

**Comment by aabewhite** [comment]
Sorry, they are in fact supported now!
https://skip.tools/docs/modules/skip-ui/#named-colors

Use File > Packages > Update to Latest Versions to make sure you have an updated skip-ui version

**Reply by tomauty** [reply]
Maybe I'm doing something wrong...here's what I have:
```swift
  static let red400 = Color("R400", bundle: .module)
```

<img width="793" alt="image" src="https://github.com/user-attachments/assets/51f4f8fd-072c-4419-b16b-b02fb067eb67" />

```swift

  public var body: some View {
    ZStack {
      Color.DarkPrimary.ignoresSafeArea(.all)
      VStack(alignment: .leading) {
        Spacer()
        Text("Title3").title3Font()
          .foregroundStyle(Color.red400)
        Text("Title4").title4Font()
          .foregroundStyle(Color.red400)
        Spacer()
        Text("heading5").heading5Font()
        Text("heading4").heading4Font()
        Text("heading3").heading3Font()
        Text("heading2").heading2Font()
        Text("heading1").heading1Font()
        Text("subheadline").subheadlineFont()
        Spacer()
        Text("body").bodyFont()
        Text("body2").body2Font()
        Text("body3").body3Font()
        Spacer()
        Text("footnote").footnoteFont()
        Text("caption3").caption3Font()
        Text("caption2").caption2Font()
        Text("caption1").caption1Font()
        Spacer()
//        Button(action: { }) { Text("Base Button") }.primaryButtonStyle()
//        Button(action: { }) { Text("Secondary Button") }.secondaryButtonStyle()
//        Button(action: { }) { Text("Tertiary Button") }.tertiaryButtonStyle()
      }
      .foregroundStyle(Color.grey25)
    }
  }
```

<img width="794" alt="image" src="https://github.com/user-attachments/assets/0e868a93-54ae-4b6b-a7e7-75dd988cc3fb" />


If I remove the foregroundStyle entries the text does show on Android, but a semi-dark-grey. Hopefully this is just an obvious user error on my part, but appreciate any advice. Updated all packages to latest. Ty!

**Reply by aabewhite** [reply]
Hmm... is your R400.colorset in your Module.xcassets?
![Screenshot 2025-01-30 at 2 31 18 PM](https://github.com/user-attachments/assets/0960b086-72d9-453d-95d2-7b40ecd46344)

**Reply by tomauty** [reply]
Yep! 
<img width="573" alt="image" src="https://github.com/user-attachments/assets/2ec927d6-6fbc-4011-9851-53b363a31741" />
Sorry had to reupload

**Reply by aabewhite** [reply]
I can't tell from the image - is it in a folder within the xcassets or at the root level? We have reports that folders may not work

**Reply by tomauty** [reply]
It's at the root level. I think I was one of the folder reporters actually, haha. Tried folders earlier and got not found errors. No errors here just no colors.

**Reply by aabewhite** [reply]
Hmm.... 

Have you tried to delete DerivedData and do a clean build? I know that's a pain, but I can't think of what the issue is. We can see named colors working in our Showcase sample, and there isn't a difference as far as I can tell. Is there anything different about how the color is specified in the info panel?
![Screenshot 2025-01-30 at 2 38 48 PM](https://github.com/user-attachments/assets/8a3a84b4-4c36-4c6c-8d1c-5e960c131e6c)

**Reply by aabewhite** [reply]
Actually if you send me R400.colorset/Contents.json, that might tell me if there's anything in the spec that we're not handling

**Reply by tomauty** [reply]
So there's nothing different. I removed DerivedData and still had the same issue. Still trying some different things over here. Thanks for the attention!

**Reply by tomauty** [reply]
Here's the json for R400.colorset/Contents.json
```json
{
  "colors" : [
    {
      "color" : {
        "color-space" : "srgb",
        "components" : {
          "alpha" : "1.000",
          "blue" : "0x82",
          "green" : "0x6E",
          "red" : "0xF3"
        }
      },
      "idiom" : "universal"
    },
    {
      "appearances" : [
        {
          "appearance" : "luminosity",
          "value" : "dark"
        }
      ],
      "color" : {
        "color-space" : "srgb",
        "components" : {
          "alpha" : "1.000",
          "blue" : "0xFF",
          "green" : "0xFF",
          "red" : "0xFE"
        }
      },
      "idiom" : "universal"
    }
  ],
  "info" : {
    "author" : "xcode",
    "version" : 1
  }
}
```

edit: ignore my opacity comment. Couldn't tell the difference between working and broken colors haha

**Reply by aabewhite** [reply]
Oh... my guess is the hex values. Can you try one that uses all decimals and no hex?

**Reply by tomauty** [reply]
Yep that was it. You're my savior. 
<img width="838" alt="image" src="https://github.com/user-attachments/assets/85619f8d-e268-4443-849d-d9f3e7c4d827" />
Now have to go and convert a lotta colors :D

**Reply by tomauty** [reply]
Apparently just changing the input method should convert it. 
<img width="265" alt="image" src="https://github.com/user-attachments/assets/7ab14fce-a4e6-4b5e-81c8-5940ea2ad25f" />
For anyone encountering the same issue. Thank god I don't have to re-copy all these values.

**Reply by aabewhite** [reply]
I'll see about supporting hex in the future, but not sure exactly when. I just didn't realize that was a thing! Thanks for helping to debug

**Reply by tomauty** [reply]
Yeah I chose that because it was easiest to copy/paste from a Figma diagram haha. You can actually select everything in the catalog, then cmd-select each color, then set the input method for all of them all at once to convert over, so it wasn't too painful.
